Introduce a use case-driven section on declaring repositories in the user guide

bmuschko opened the pull request:
### Context

This is the initial draft of a use case-driven section on declaring repositories for the user guide. 

- Reference material was moved into the glossary.
- Reference material was trimmed down and polished.
- New content was added based on the proposed outline. 

The use cases I put out are the ones which I believe users encounter a lot. We might want to work on the titles and the use case a bit more. Happy to hear your feedback.

The reference material uses a table. Should we turn this into a list?

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagement.adoc:
```diff
@@ -370,490 +370,64 @@ Once a configuration is resolved it is immutable. Changing its state or the stat
-[[sec:repositories]]
-=== Repositories
+[[sec:declaring_repositories]]
+=== Declaring repositories
```
Should we consider "Declaring Repositories" as a separate chapter? I think it would be _much_ more discoverable than buried in the Advanced Dependency Management chapter.

bmuschko replied:
In the long run we'll definitely want to break up the whole dependency management chapter. I am just wondering when it is a good time to do so. Optimally that would be the case if we have a consistent outline/structure in the current chapter. Imagine we'd just break out "declaring dependencies" and "declaring repositories" right now then I'd think that the rest would get rather inconsistent. Maybe we can make the overall flow better and then break it up. Would you agree?

eriwen replied:
I would recommend we prioritize discoverability over "flow" because users do not read chapters and sections in order... but you're in charge of this chapter so I'll let you make the call here, as long as we get this settled by Gradle 4.6 I think.

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagement.adoc:
```diff
@@ -370,490 +370,64 @@ Once a configuration is resolved it is immutable. Changing its state or the stat
 ++++
 
-An overview of which Resolvers are offered by Ivy and thus also by Gradle can be found http://ant.apache.org/ivy/history/latest-milestone/settings/resolvers.html[here]. With Gradle you just don't configure them via XML but directly via their API.
+The order of declaration determines how Gradle will check for dependencies at runtime. You can learn more about <<sec:dependency_resolution,Gradle's resolution mechanism>> below.
```
This is a really important point, so I think we should highlight it somehow, perhaps with an admonition.

eriwen replied:
:o: Avoid positional descriptions like "below"

bmuschko replied:
I knew that you'd complain about "below" but I couldn't come up with a good alternative. Leaving out "below" makes the sentence sound kind of weird. Generally speaking, I don't mind positional description. Yes, it might need to change if things are shuffled around but from my perspective that's the job of the person changing the text + the review to ensure that those positional descriptions still work.

bmuschko replied:
RE: admonition - done. BTW: Do we support other admonitions than `NOTE`? Thinking about `WARNING`, `IMPORTANT`, `TIP` etc.

eriwen replied:
There are

> note, caution, warning, tip, and important

according to http://www.sagehill.net/docbookxsl/CustomAdmonitions.html

eriwen commented on subprojects/docs/src/docs/userguide/dependencyManagement.adoc:
```diff
@@ -370,490 +370,64 @@ Once a configuration is resolved it is immutable. Changing its state or the stat
-==== Maven Google repository
-
-The Google repository hosts Android-specific artifacts including the Android SDK. For usage examples please the [relevant documentation](https://developer.android.com/studio/build/dependencies.html#google-maven).
++++++
```
Totally optional nitpick: this could be a bit wider.

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/repositoryTypes.adoc:
```diff
@@ -0,0 +1,390 @@
+++++
+
+[[sub:maven_google]]
+=== Maven Google repository
```
:o: I think it's named "Google Maven Repository" - https://developer.android.com/studio/build/dependencies.html#google-maven

bmuschko replied:
fixed

eriwen commented on subprojects/docs/src/docs/userguide/repositoryTypes.adoc:
```diff
@@ -0,0 +1,390 @@
+++++
+
+[[sub:maven_jcenter]]
+=== Maven JCenter repository
```
Probably "JCenter Maven repository" to match the paragraph below

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/repositoryTypes.adoc:
```diff
@@ -0,0 +1,390 @@
+
+The following system properties can be used to configure the interactions with s3 repositories:
+
+.S3 Configuration Properties
```
Should be "S3 configuration properties" for sentence case

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/repositoryTypes.adoc:
```diff
@@ -0,0 +1,390 @@
+
+The following system properties can be used to configure the interactions with Google Cloud Storage repositories:
+
+.Google Cloud Storage Configuration Properties
```
Sentence case

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/repositoryTypes.adoc:
```diff
@@ -0,0 +1,390 @@
+[[sub:gcs_configuration_properties]]
+==== Google Cloud Storage configuration properties
+
+The following system properties can be used to configure the interactions with Google Cloud Storage repositories:
```
Link to Google Cloud Storage maybe.

bmuschko replied:
done

eriwen commented on subprojects/docs/src/docs/userguide/repositoryTypes.adoc:
```diff
@@ -0,0 +1,390 @@
+[[repository_types]]
+== Repository types
```
Would you prefer sentence case or title case for titles here?

bmuschko replied:
I don't mind using sentence case consistently across the board.

eriwen commented on subprojects/docs/src/docs/userguide/repositoryTypes.adoc:
```diff
@@ -0,0 +1,390 @@
+[[repository_types]]
```
This is an excellent chapter.

bmuschko replied:
Thanks!

bmuschko commented on subprojects/docs/src/samples/userguide/repositories/declaringMultipleRepositories/build.gradle:
```diff
@@ -0,0 +1,24 @@
+}
+
+dependencies {
+    libs 'com.restlet.client:commons:2.0.0'
```
Won't be rendered in the docs but changed it to double quotes. I think we should standardize _a_ case across the whole documentation. My personal preference for Groovy is:

- Single quotes by default.
- Double quotes only for String interpolation.

bmuschko commented:
> I think we've really got to split it out from the big Dependency Management chapter so users can find it.

Let's do it in a separate PR. Please also have a look at my [other comment](https://github.com/gradle/gradle/pull/3910#discussion_r158850246).

> Most everything else is a nitpick.

Addressed your comment. Could you have another look?

bmuschko commented:
> There are note, caution, warning, tip, and important according to http://www.sagehill.net/docbookxsl/CustomAdmonitions.html.

I believe I tried them but they didn't render properly with our Docbook implementation.

eriwen commented:
Regarding admonitions, I can help with styling if you want to use other types
